src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -274,28 +456,76 @@ void CoreWorkerDirectTaskSubmitter::RequestNewWorkerIfNeeded(
   }
 
   // Check whether we really need a new worker or whether we have
-  // enough room in an existing worker's pipeline to send the new tasks
-  if (!scheduling_key_entry.AllPipelinesToWorkersFull(max_tasks_in_flight_per_worker_)) {
+  // enough room in an existing worker's pipeline to send the new tasks. If the pipelines
+  // are not full, we do not request a new worker (unless work stealing is enabled, in
+  // which case we can request a worker under the Eager Worker Requesting mode)
+  if (!scheduling_key_entry.AllPipelinesToWorkersFull(max_tasks_in_flight_per_worker_) &&
+      !work_stealing_) {
     // The pipelines to the current workers are not full yet, so we don't need more
     // workers.
     return;
   }
 
-  TaskSpecification &resource_spec = task_queue.front();
+  auto &task_queue = scheduling_key_entry.task_queue;
+  // Check if the task queue is empty. If that is the case, it only makes sense to
+  // consider requesting a new worker if work stealing is enabled, and there is at least a
+  // worker with stealable tasks
+  if (task_queue.empty()) {
+    if (!work_stealing_) {
+      if (scheduling_key_entry.CanDelete()) {
+        // We can safely remove the entry keyed by scheduling_key from the
+        // scheduling_key_entries_ hashmap.
+        scheduling_key_entries_.erase(scheduling_key);
+      }
+      return;
+    }
+
+    // We don't have any of this type of task to run.
+    bool found = false;
+
+    // Look for a worker (with the current scheduling key) with tasks to steal. If we
```

Comment:
I think we could simplify this by keeping the total sum of in-flight tasks and dividing that by the number of workers (including thieves). That way, we don't need to loop through the workers each time we need to submit a task.

Reply:
just changed this here: https://github.com/ray-project/ray/pull/13570/commits/42ff5e73937278f5ab9f17e1c27d6ff2339159a3